File: android/src/main/java/com/metrickit/SafeExecutor.kt
package com.metrickit

import android.util.Log

/**
 * Crash-free guard (Android side). Wraps reads so any failure is logged and
 * swallowed rather than thrown into React Native. See ARCHITECTURE.md.
 */
internal object SafeExecutor {
  private const val TAG = "react-native-metrickit"

  fun attempt(context: String, block: () -> Unit) {
    try {
      block()
    } catch (t: Throwable) {
      Log.w(TAG, "$context failed (ignored)", t)
    }
  }

  fun <T> value(context: String, fallback: T, block: () -> T): T {
    return try {
      block()
    } catch (t: Throwable) {
      Log.w(TAG, "$context failed (ignored)", t)
      fallback
    }
  }
}

File: ios/SafeExecutor.swift
import Foundation

/// Crash-free guard (Swift side). Wraps throwing work so a parse failure is
/// logged and swallowed rather than propagated. See ARCHITECTURE.md.
enum SafeExecutor {
  static func attempt(_ context: String, _ block: () throws -> Void) {
    do {
      try block()
    } catch {
      NSLog("[react-native-metrickit] %@ failed (ignored): %@", context, String(describing: error))
    }
  }

  static func value<T>(_ context: String, default fallback: T, _ block: () throws -> T) -> T {
    do {
      return try block()
    } catch {
      NSLog("[react-native-metrickit] %@ failed (ignored): %@", context, String(describing: error))
      return fallback
    }
  }
}
